File: leetcode/maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.cpp
/**************************************
* @author dujiong
* @date 2016.9.17
* @version V0.1
**************************************/

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode *root) {
        if(root == NULL){
        	return 0;
        }
        int ret=1;
        if(root->left!=NULL || root->right!=NULL){
            ret+=std::max(maxDepth(root->left),maxDepth(root->right));
        }
        return ret;
    }
};

File: leetcode-golang/Maximum Depth of Binary Tree/Maximum Depth of Binary Tree.go
type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

//bingo
func maxDepth(root *TreeNode) int {
    if root == nil {
		return 0
	}
	left, right := 1, 1
	left += maxDepth(root.Left)
	right += maxDepth(root.Right)
	return int(math.Max(float64(left), float64(right)))
}